### PR TITLE
Feature/many to many support

### PR DESCRIPTION
**Pull request by pietaj**
I am targeting this branch, because the bundle does not support m:n auditing and the subject was already mentioned 
here https://github.com/sonata-project/EntityAuditBundle/issues/302, and here https://github.com/sonata-project/EntityAuditBundle/pull/321 but was never finished

## Changelog

```markdown
### Added
- Support for ManyToMany
```

### CODE REVIEW

**Comment by VincentLanglet** [comment]
I fixed the cs in https://github.com/sonata-project/EntityAuditBundle/pull/510.
Please rebase and revert the extra style changes.

We don't work on this library a lot so the more the PR will be small, the easier it will be for us to review ^^

And take a look at the build ;)

**Comment by pietaj** [comment]
I've rebased the code and reverted the style changes.

I've also left the deprecation fixes out of this PR as it makes the code harder to review.

**Comment by pietaj** [comment]
@VincentLanglet I've checked the stan / psalm issues reported by the CI and I can fix them, no problem, but this will make the code changes in this PR even more complex, as there are quite a lot of things to review already. Shall I proceed with the fixes and throw them all in a separate commit?  What would be the most convenient way to proceed?

File: src/AuditReader.php [hunk 1]
```diff
@@ -984,6 +984,134 @@ private function createEntity($className, array $columnMap, array $data, $revisi
...
+                        $values[] = $classMetadata->reflFields['id']->getValue($entity);
+                    }
+
+                    $whereSQL = implode(' AND ', $whereId);
+                    $columnList = [
+                        $this->config->getRevisionFieldName(),
+                        $this->config->getRevisionTypeFieldName(),
+                    ];
+                    $tableName = $this->config->getTablePrefix(
+                        ).$assoc['joinTable']['name'].$this->config->getTableSuffix();
```
Comment (VincentLanglet) [inline src/AuditReader.php]:
```suggestion
                    $tableName = $this->config->getTablePrefix()
                        .$assoc['joinTable']['name']
                        .$this->config->getTableSuffix();
```

File: src/AuditReader.php [hunk 2]
```diff
@@ -984,6 +984,134 @@ private function createEntity($className, array $columnMap, array $data, $revisi
...
+
+                    foreach ($assoc['relationToTargetKeyColumns'] as $targetKeyJoinColumn => $targetKeyColumn) {
+                        $columnList[] = $targetKeyJoinColumn;
+                    }
+                    $query = 'SELECT '.implode(
+                            ', ',
+                            $columnList
+                        ).' FROM '.$tableName.' e WHERE '.$whereSQL.' ORDER BY e.'.$this->config->getRevisionFieldName(
+                        ).' DESC';
```
Comment (VincentLanglet) [inline src/AuditReader.php]: Prefer using sprintf

File: src/AuditReader.php [hunk 3]
```diff
@@ -984,6 +984,134 @@ private function createEntity($className, array $columnMap, array $data, $revisi
...
+                        }
+
+                        $whereSQL = implode(' AND ', $whereId);
+                        $columnList = [
+                            $this->config->getRevisionFieldName(),
+                            $this->config->getRevisionTypeFieldName(),
+                        ];
+
+                        $tableName = $this->config->getTablePrefix(
+                            ).$targetAssoc['joinTable']['name'].$this->config->getTableSuffix();
```
Comment (VincentLanglet) [inline src/AuditReader.php]:
```suggestion
                        $tableName = $this->config->getTablePrefix()
                            .$targetAssoc['joinTable']['name']
                            .$this->config->getTableSuffix();
```

File: src/AuditReader.php [hunk 4]
```diff
@@ -984,6 +984,134 @@ private function createEntity($className, array $columnMap, array $data, $revisi
...
+
+                        foreach ($targetAssoc['relationToSourceKeyColumns'] as $sourceKeyJoinColumn => $sourceKeyColumn) {
+                            $columnList[] = $sourceKeyJoinColumn;
+                        }
+                        $query = 'SELECT '.implode(
+                                ', ',
+                                $columnList
+                            ).' FROM '.$tableName.' e WHERE '.$whereSQL.' ORDER BY e.'.$this->config->getRevisionFieldName(
+                            ).' DESC';
```
Comment (VincentLanglet) [inline src/AuditReader.php]: prefer sprintf

File: src/EventListener/LogRevisionsListener.php [hunk 5]
```diff
@@ -389,6 +426,49 @@ private function getInsertRevisionSQL(ClassMetadata $class): string
...
+     *
+     * @return string
+     *
+     * @psalm-param ClassMetadata<object> $targetClass
+     * @psalm-param array{cache?: array, cascade: array<string>, declared?: class-string, fetch: mixed, fieldName: string, id?: bool, inherited?: class-string, indexBy?: string, inversedBy: null|string, isCascadeRemove: bool, isCascadePersist: bool, isCascadeRefresh: bool, isCascadeMerge: bool, isCascadeDetach: bool, isOnDeleteCascade?: bool, isOwningSide: true, joinColumns?: array<array{name: string, referencedColumnName: string, unique?: bool, quoted?: bool, fieldName?: string, onDelete?: string, columnDefinition?: string, nullable?: bool}>, joinColumnFieldNames?: array<string, string>, joinTable?: array, joinTableColumns?: list<mixed>, mappedBy: null|string, orderBy?: array, originalClass?: class-string, originalField?: string, orphanRemoval?: bool, relationToSourceKeyColumns?: array, relationToTargetKeyColumns?: array, sourceEntity: class-string, sourceToTargetKeyColumns?: array<string, string>, targetEntity: class-string, targetToSourceKeyColumns?: array<string, string>, type: int, unique?: bool} $assoc
```
Comment (VincentLanglet) [inline src/EventListener/LogRevisionsListener.php]: You can use @phpstan-import-type AssociationMapping from ... https://github.com/doctrine/orm/blob/2.13.x/lib/Doctrine/ORM/Mapping/ClassMetadataInfo.php#L110
instead on the class level.

And then use it in the phpdoc.

File: src/EventListener/LogRevisionsListener.php [hunk 6]
```diff
@@ -389,6 +426,49 @@ private function getInsertRevisionSQL(ClassMetadata $class): string
         return $this->insertRevisionSQL[$class->name];
     }
 
+    /**
+     * @param ClassMetadata $class
```
Comment (VincentLanglet) [inline src/EventListener/LogRevisionsListener.php]:
```suggestion
     * @param ClassMetadata<object> $class
```
avoid the need of psalm-param

File: src/EventListener/LogRevisionsListener.php [hunk 7]
```diff
@@ -389,6 +426,49 @@ private function getInsertRevisionSQL(ClassMetadata $class): string
...
+     * @param $targetClass
+     * @param $assoc
+     * @param (((bool|string)[]|mixed|string)[]|bool|int|mixed|null|string)[] $assoc
+     *
+     * @return string
+     *
+     * @psalm-param ClassMetadata<object> $targetClass
```
Comment (VincentLanglet) [inline src/EventListener/LogRevisionsListener.php]: We prefer phpstan-param

File: src/EventListener/LogRevisionsListener.php [hunk 8]
```diff
@@ -403,34 +483,51 @@ private function saveRevisionEntityData(ClassMetadata $class, array $entityData,
             if ($class->isInheritanceTypeJoined() && $class->isInheritedAssociation($field)) {
                 continue;
             }
-            if (
-                ($assoc['type'] & ClassMetadata::TO_ONE) === 0
+            if (!(($assoc['type'] & ClassMetadata::TO_ONE) === 0
```
Comment (VincentLanglet) [inline src/EventListener/LogRevisionsListener.php]: Change !($a || $b ||$c) to !a && !b && !c

**Review by VincentLanglet — CHANGES_REQUESTED** [review]

**Comment by VincentLanglet** [comment]
> @VincentLanglet I've checked the stan / psalm issues reported by the CI and I can fix them, no problem, but this will make the code changes in this PR even more complex, as there are quite a lot of things to review already. Shall I proceed with the fixes and throw them all in a separate commit? What would be the most convenient way to proceed?

Yes please fix the issues. 

Lot of them where already fix on some others places of the code with some if checks or some assert. You can do the same.

**Comment by pietaj** [comment]
> > @VincentLanglet I've checked the stan / psalm issues reported by the CI and I can fix them, no problem, but this will make the code changes in this PR even more complex, as there are quite a lot of things to review already. Shall I proceed with the fixes and throw them all in a separate commit? What would be the most convenient way to proceed?
> 
> Yes please fix the issues.
> 
> Lot of them where already fix on some others places of the code with some if checks or some assert. You can do the same.

Oh my, thanks for the ultra fast reply, I didn't mean to push ;) I will fix all issues and will come back to you after everything will pass green

**Comment by VincentLanglet** [comment]
Any issue @pietaj ? :)

**Comment by pietaj** [comment]
My only issue was my day job :) did the requested psalm / phpstan compliance cleanup and the requested changes

**Comment by VincentLanglet** [comment]
> My only issue was my day job :) did the requested psalm / phpstan compliance cleanup and the requested changes

Seems like build is still failing

**Comment by pietaj** [comment]
That's odd ... both rector and phpstan pass on dev machine. I'll have a closer look

**Comment by VincentLanglet** [comment]
> That's odd ... both rector and phpstan pass on dev machine. I'll have a closer look

Tru to update your dependencies maybe ?

**Comment by pietaj** [comment]
yeah, thanks for the tip. Outdated dependencies, I'll rerun and fix the rest

**Comment by pietaj** [comment]
@VincentLanglet one of the failing checks is linked with rector, failing with

`PHP Fatal error:  Uncaught _PHPStan_9a6ded56a\Nette\DI\InvalidConfigurationException: The item 'conditionalTags › PHPStan\Rules\DisallowedConstructs\DisallowedLooseComparisonRule › phpstan.rules.rule' expects to be bool, array given. in phar:///home/runner/work/EntityAuditBundle/EntityAuditBundle/vendor/rector/rector/vendor/phpstan/phpstan/phpstan.phar/vendor/nette/di/src/DI/Compiler.php:219`

I did some quick checks and the error is linked with phpstan/phpstan bumping from `1.8.5 => 1.8.6`. Did you encounter a similar problems in other sonata projects as well?

**Comment by jordisala1991** [comment]
> @VincentLanglet one of the failing checks is linked with rector, failing with
> 
> `PHP Fatal error: Uncaught _PHPStan_9a6ded56a\Nette\DI\InvalidConfigurationException: The item 'conditionalTags › PHPStan\Rules\DisallowedConstructs\DisallowedLooseComparisonRule › phpstan.rules.rule' expects to be bool, array given. in phar:///home/runner/work/EntityAuditBundle/EntityAuditBundle/vendor/rector/rector/vendor/phpstan/phpstan/phpstan.phar/vendor/nette/di/src/DI/Compiler.php:219`
> 
> I did some quick checks and the error is linked with phpstan/phpstan bumping from `1.8.5 => 1.8.6`. Did you encounter a similar problems in other sonata projects as well?

Yes, ignore rector for now.

FYI, error already reported: https://github.com/rectorphp/rector/issues/7509, waiting for a fix. (The most you can do is upgrade to rector ^0.14, to ensure we will get the fix.

File: src/AuditReader.php [hunk 9]
```diff
@@ -237,7 +237,9 @@ public function find($className, $id, $revision, array $options = [])
         $columnMap = [];
 
         foreach ($classMetadata->fieldNames as $columnName => $field) {
-            $tableAlias = $classMetadata->isInheritanceTypeJoined() && $classMetadata->isInheritedField($field) && !$classMetadata->isIdentifier($field)
+            $tableAlias = $classMetadata->isInheritanceTypeJoined() && $classMetadata->isInheritedField(
+                $field
+            ) && !$classMetadata->isIdentifier($field)
```
Comment (VincentLanglet) [inline src/AuditReader.php]: This reformatting is weird. I would prefer
```suggestion
            $tableAlias = $classMetadata->isInheritanceTypeJoined()
                && $classMetadata->isInheritedField($field)
                && !$classMetadata->isIdentifier($field)
```

File: src/AuditReader.php [hunk 10]
```diff
@@ -404,7 +417,9 @@ public function findEntitiesChangedAtRevision($revision)
 
             foreach ($classMetadata->fieldNames as $columnName => $field) {
                 $type = Type::getType($classMetadata->fieldMappings[$field]['type']);
-                $tableAlias = $classMetadata->isInheritanceTypeJoined() && $classMetadata->isInheritedField($field) && !$classMetadata->isIdentifier($field)
+                $tableAlias = $classMetadata->isInheritanceTypeJoined() && $classMetadata->isInheritedField(
+                    $field
+                ) && !$classMetadata->isIdentifier($field)
```
Comment (VincentLanglet) [inline src/AuditReader.php]:
```suggestion
                $tableAlias = $classMetadata->isInheritanceTypeJoined()
                    && $classMetadata->isInheritedField($field)
                    && !$classMetadata->isIdentifier($field)
```

File: src/AuditReader.php [hunk 11]
```diff
@@ -541,7 +559,8 @@ public function findRevisions($className, $id)
         }
 
         $query = 'SELECT r.* FROM '.$this->config->getRevisionTableName().' r '.
-                 'INNER JOIN '.$tableName.' e ON r.id = e.'.$this->config->getRevisionFieldName().' WHERE '.$whereSQL.' ORDER BY r.id DESC';
+            'INNER JOIN '.$tableName.' e ON r.id = e.'.$this->config->getRevisionFieldName(
+            ).' WHERE '.$whereSQL.' ORDER BY r.id DESC';
```
Comment (VincentLanglet) [inline src/AuditReader.php]: this formatting is weird

File: src/AuditReader.php [hunk 12]
```diff
@@ -889,7 +915,9 @@ private function createEntity($className, array $columnMap, array $data, $revisi
                                 $pf[$foreign] = $key;
                             }
                         } elseif (null !== $mappedBy) {
-                            $otherEntityAssoc = $this->em->getClassMetadata($targetEntity)->associationMappings[$mappedBy];
+                            $otherEntityAssoc = $this->em->getClassMetadata(
+                                $targetEntity
+                            )->associationMappings[$mappedBy];
```
Comment (VincentLanglet) [inline src/AuditReader.php]: this formatting is weird

File: src/EventListener/LogRevisionsListener.php [hunk 13]
```diff
@@ -389,6 +426,45 @@ private function getInsertRevisionSQL(ClassMetadata $class): string
         return $this->insertRevisionSQL[$class->name];
     }
 
+    /**
+     * @param ClassMetadata        $class
+     * @param array<string, mixed> $assoc
+     *
+     * @psalm-param ClassMetadata<object> $targetClass
```
Comment (VincentLanglet) [inline src/EventListener/LogRevisionsListener.php]:
```suggestion
     * @phpstan-param ClassMetadata<object> $targetClass
```

File: src/EventListener/LogRevisionsListener.php [hunk 14]
```diff
@@ -389,6 +426,45 @@ private function getInsertRevisionSQL(ClassMetadata $class): string
...
+     */
+    private function getInsertJoinTableRevisionSQL($class, ClassMetadata $targetClass, array $assoc): string
+    {
+        $cacheKey = $class->name.'.'.$targetClass->name;
+        if (!isset($this->insertJoinTableRevisionSQL[$cacheKey])
+            && isset($assoc['relationToSourceKeyColumns'], $assoc['relationToTargetKeyColumns'], $assoc['joinTable']['name'])) {
+            $placeholders = ['?', '?'];
+
+            // $tableName = $this->config->getTableName($class);
```
Comment (VincentLanglet) [inline src/EventListener/LogRevisionsListener.php]: Why adding a code commented ? can we remove it ?

**Review by VincentLanglet — CHANGES_REQUESTED** [review]
It's ok for me, except some formatting issues.
Is this PR ok for you @jordisala1991 ?

File: tests/RelationTest.php [hunk 15]
```diff
@@ -306,8 +307,7 @@ public function testOneToOne(): void
     }
 
     /**
-     * This test verifies the temporary behaviour of audited entities with M-M relationships
-     * until https://github.com/simplethings/EntityAudit/issues/85 is implemented.
+     * This test verifies the temporary behaviour of audited entities with M-M relationships.
```
Comment (jordisala1991) [inline tests/RelationTest.php]: I think this comment is outdated. The behavior implemented right now is not temporary right?

**Review by jordisala1991 — COMMENTED** [review]
Do we need more information on the readme, or should we only remove the last todo?

**Comment by jordisala1991** [comment]
This Pr also needs a changelog.

**Review by VincentLanglet — DISMISSED** [review]

**Review by VincentLanglet — APPROVED** [review]

**Review by jordisala1991 — APPROVED** [review]

**Comment by jordisala1991** [comment]
Thanks @pietaj

**Comment by jordisala1991** [comment]
I guess the todo list on the readme should be updated too.

**Pull request merged**